Add provisioning logic

dhellmann opened the pull request:
This successfully triggers provisioning.

dtantsur commented on deploy/crds/worker-0.yaml:
```diff
@@ -19,3 +19,6 @@ spec:
   bootMACAddress: 00:73:49:3a:76:8e
+  image:
+    url: "http://172.22.0.1/images/redhat-coreos-maipo-latest.qcow2"
+    checksum: "97830b21ed272a3d854615beb54cf004"
```
btw we understand URLs here starting with Stein IPA images.

stbenjam commented on pkg/provisioner/ironic/ironic.go:
```diff
@@ -390,6 +383,165 @@ func (p *ironicProvisioner) InspectHardware() (result provisioner.Result, err er
+	// Wait for the host to become available
+	if p.status.State == stateMakingAvailable {
+		if ironicNode.ProvisionState != nodes.Available {
+			p.log.Info("waiting for host to become available")
```
What happens if a host never becomes available? For example, Ironic goes to `clean failed` - does this mean it'll just return dirty forever?

dhellmann replied:
Yeah, for this first pass I'm just trying to get 1 happy path working. There are going to be lots of error handling updates to make.

dhellmann commented:
I rebased this. The latest series of patches allows me to write the image to the VM, and the VM boots. I would like to merge this before working on some of the error handling cases.

russellb commented:
I'm fine with this based on existing reviews.  @dhellmann I'll leave the merge to you so you can decide how you want to squash this

dhellmann commented:
I rebased and squashed a couple of the fix-up patches into the main provisioning patch, but I'll leave the others as they are because they're somewhat unrelated changes.
